File: synchronization/src/main/java/de/cyface/synchronization/SyncService.java
package de.cyface.synchronization;

import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.IBinder;
import android.preference.PreferenceManager;

import androidx.annotation.NonNull;

import de.cyface.uploader.DefaultAuthenticator;
import de.cyface.uploader.DefaultUploader;
import de.cyface.utils.Validate;

/**
 * The synchronisation <code>Service</code> used to bind the synchronisation adapter to the Android framework.
 * <p>
 * Further details are described in the <a href=
 * "https://developer.android.com/training/sync-adapters/creating-sync-adapter.html#CreateSyncAdapterService">Android
 * documentation</a>.
 *
 * @author Armin Schnabel
 * @author Klemens Muthmann
 * @version 1.0.8
 * @since 2.0.0
 */
public final class SyncService extends Service {

    /**
     * Logging TAG to identify logs associated with the {@link WiFiSurveyor}.
     */
    @SuppressWarnings({"FieldCanBeLocal", "unused"}) // we add and move logs often, so keep it
    public static final String TAG = Constants.TAG + ".syncSvc";
    /**
     * The settings key used to identify the settings storing the URL of the server to upload data to.
     */
    public static final String SYNC_ENDPOINT_URL_SETTINGS_KEY = "de.cyface.sync.endpoint";
    /**
     * The settings key used to identify the settings storing the OAuth configuration JsonObject as String.
     */
    public static final String OAUTH_CONFIG_SETTINGS_KEY = "de.cyface.oauth.config";
    /**
     * The synchronisation adapter this service is supposed to call.
     * <p>
     * Singleton isn't what they call a beauty. Nevertheless this is how it is specified in the documentation. Maybe try
     * to change this after it runs.
     */
    private static SyncAdapter syncAdapter = null;
    /**
     * Lock object used to synchronize synchronisation adapter creation as described in the Android documentation.
     */
    private static final Object LOCK = new Object();

    @Override
    public void onCreate() {
        synchronized (LOCK) {
            if (syncAdapter == null) {
                final var oauth = new OAuth2(getApplicationContext()); // move to DefaultAuthenticator
                final var authApi = authApi(getApplicationContext());
                final var collectorApi = collectorApi(getApplicationContext());
                syncAdapter = new SyncAdapter(getApplicationContext(), true, new DefaultAuthenticator(authApi),
                        new DefaultUploader(collectorApi), oauth);
            }
        }
    }

    @Override
    public IBinder onBind(Intent intent) {
        return syncAdapter.getSyncAdapterBinder();
    }

    /**
     * Reads the Collector API URL from the preferences.
     *
     * @param context The `Context` required to read the preferences
     * @return The URL as string
     */
    private String collectorApi(@NonNull final Context context) {
        final var preferences = PreferenceManager.getDefaultSharedPreferences(context);
        final var apiEndpoint = preferences.getString(SyncService.SYNC_ENDPOINT_URL_SETTINGS_KEY, null);
        Validate.notNull(
                apiEndpoint,
                "Sync canceled: Server url not available. Please set the applications server url preference.");
        return apiEndpoint;
    }

    /**
     * Reads the Auth URL from the preferences.
     *
     * @param context The `Context` required to read the preferences
     * @return The URL as string
     */
    private String authApi(@NonNull final Context context) {
        final var preferences = PreferenceManager.getDefaultSharedPreferences(context);
        final var apiEndpoint = preferences.getString(OAUTH_CONFIG_SETTINGS_KEY, null);
        Validate.notNull(
                apiEndpoint,
                "Sync canceled: Auth url not available. Please set the applications server url preference.");
        return apiEndpoint;
    }
}

File: synchronization/src/cyface/kotlin/de/cyface/synchronization/SyncService.kt
/*
 * Copyright 2017-2023 Cyface GmbH
 *
 * This file is part of the Cyface SDK for Android.
 *
 * The Cyface SDK for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The Cyface SDK for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the Cyface SDK for Android. If not, see <http://www.gnu.org/licenses/>.
 */
package de.cyface.synchronization

import android.app.Service
import android.content.Intent
import android.os.IBinder
import de.cyface.uploader.DefaultUploader
import de.cyface.utils.Validate
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.runBlocking

/**
 * The synchronisation `Service` used to bind the synchronisation adapter to the Android framework.
 *
 * Further details are described in the [Android
 * documentation](https://developer.android.com/training/sync-adapters/creating-sync-adapter.html#CreateSyncAdapterService).
 *
 * @author Armin Schnabel
 * @author Klemens Muthmann
 * @version 1.0.9
 * @since 2.0.0
 */
class SyncService : Service() {
    override fun onCreate() {
        synchronized(LOCK) {
            if (syncAdapter == null) {
                val collectorApi = collectorApi()
                syncAdapter = SyncAdapter(
                    applicationContext,
                    true,
                    OAuth2(applicationContext, CyfaceAuthenticator.settings),
                    DefaultUploader(collectorApi),
                )
            }
        }
    }

    override fun onBind(intent: Intent): IBinder? {
        return syncAdapter!!.syncAdapterBinder
    }

    /**
     * Reads the Collector API URL from the preferences.
     *
     * @return The URL as string
     */
    private fun collectorApi(): String {
        val apiEndpoint =
            runBlocking { CyfaceAuthenticator.settings.collectorUrlFlow.first() }
        Validate.notNull(
            apiEndpoint,
            "Sync canceled: Server url not available. Please set the applications server url preference."
        )
        return apiEndpoint
    }

    companion object {
        /**
         * The synchronisation adapter this service is supposed to call.
         *
         * Singleton isn't what they call a beauty. Nevertheless this is how it is specified in the documentation. Maybe try
         * to change this after it runs.
         */
        private var syncAdapter: SyncAdapter? = null

        /**
         * Lock object used to synchronize synchronisation adapter creation as described in the Android documentation.
         */
        private val LOCK = Any()
    }
}